### DISCUSSION TITLE
Support "Find All References" and "goto" on special methods e.g. __getitem__ and __call__

### DISCUSSION THREAD

**Discussion by lainets** [microsoft/pylance-release#2597] [Enhancement]
Would be useful if clicking "Find All References" on special methods gives the list of places where it is used. E.g. on `__getitem__` it includes all instances of using [] as a getter on an instance of that class. Right now it only works if the special method is accessed using its name, e.g. `obj.__call__(...)`.

`__getitem__`, `__setitem__`, `__call__`, `__contains__` and `__iter__` would be especially useful. `__enter__` and `__init__` would be nice too but probably not strictly necessary as they can usually be found using the class name itself (with some manual filtering).

This would be especially useful when refactoring code. E.g. when refactoring a dict into a class, finding all references to `__getitem__` etc. would be extremely helpful.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by laundmo** [comment]
There is a additional factor to this: Currently, "Find All References" on dunder/magic methods finds all implementations of that method from all classes in the project. That means in a large project, it might well discover a lot of unrelated `__str__` implementations.

repropduction:
```py
class A:
    def __str__(self) -> str:
        return "test"

class B:
    def __str__(self) -> str:
        return "test2"
```
![image](https://github.com/microsoft/pylance-release/assets/24855949/61da1eef-7c7b-41be-910a-2fe683e4b09b)

**Reply by debonte** [reply]
@laundmo, I think this is because `__str__` is defined on `object` which is the base class of both `A` and `B`. The behavior is the same for the (non-dunder) method `foo` in the class hierarchy below.

I'm not sure if this is intentional or not. If I do "Find All References" on `A.foo` it makes sense that `A1.foo` and `A2.foo` are included in the results. But if I do "Find All References" on `A2.foo`, it's not clear to me why `A1.foo` would be included in the results. @heejaechang or @erictraut may know.

```python
class A:
    def foo(self): ...

class A1(A):
    def foo(self): ...

class A2(A):
    def foo(self): ...
```

**Reply by heejaechang** [reply]
we currently don't distinguish those 2 different cases. currently, we only provide a way to find all locations of a symbols and its overrides.

the reason being
1. even if the symbol is off `base` at `editor time`, at runtime, it might actually `derived` object. so, it might cause more confusions than helping
2. we use the same engine for `rename` so you can't just change `name` of symbol in `derived` since that could cause change in code behavior.

```python
class Base:
      def foo(self): ...
class Derived(Base):
      def foo(self): ...

def method(a: Base):
     a.foo()
     
method(Derived())
```

if one changed `Derived.foo` to `foo1`, now `method` calls `Base.foo` not `Derived.foo1`

that said, we could add new features to find exact match only if we get enough votes. but, I think we should have another `enhancement` issue rather than sharing this one to track them properly.
